File: 31_next_permutation.py
class Solution:
    def nextPermutation(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        i=0
        k=len(nums)-2
        while i>=0 and nums[i]>nums[i+1]:
            i-=1
            
        if i!=-1:
            j=i+1
            while j<len(nums) and nums[i]<nums[j]:
                j+=1
            j-=1
            nums[i],nums[j]=nums[j],nums[i]
            
            #reverse the lst
            left=i+1
            right=len(nums)-1
            while left<right:
                nums[left],nums[right]=nums[right],nums[left]
                left+=1
                right-=1
        else:
            for i in range(len(nums)-1):
                for j in range(i+1,len(nums)):
                    if nums[i]>nums[j]:
                        nums[i],nums[j]=nums[j],nums[i]
                        
        return nums
        

File: 31_Next_Permutation.cpp
class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        int n=nums.size();
        if(n<=1)
            return;
        
        int pivot_ind=n-2;
        int suff_ind=n-1;
        //to get largest index for pivot
        while(pivot_ind>=0 && nums[pivot_ind]>=nums[pivot_ind+1])
        {
            --pivot_ind;
        }
        
        if(pivot_ind>=0 )
        {
            while(nums[pivot_ind]>=nums[suff_ind])
                --suff_ind;
            
            swap(nums[pivot_ind],nums[suff_ind]);
        }
      
        pivot_ind+=1;
        
        sort(nums.begin()+pivot_ind,nums.end());
        
    }
};